### PR TITLE
Added 3 Phase APC UPS Support issue #2733 & #5504

### PR DESCRIPTION
**Pull request by NerdBlender**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/) - please do NOT submit a pull request unless you have (signing the agreement in the same pull request is fine). Your commit message for signing the agreement must appear as per the docs.
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

### CODE REVIEW

**Comment by NerdBlender** [comment]
This should also work for issue number #5504

**Comment by murrant** [comment]
I actually redid all the APC code awhile ago here: https://github.com/murrant/librenms/tree/apc-cleanup
but didn't merge it because I renamed some rrd files in some places and never got back to that.

Here is the code I used with an snmpwalk to handle any number of phases, does that work for you?
```php
    $oids = snmpwalk_cache_oid($device, 'upsHighPrecOutputCurrent', array(), 'PowerNet-MIB');
    if (empty($oids)) {
        $oids = snmpwalk_cache_oid($device, 'upsAdvOutputCurrent', $oids, 'PowerNet-MIB');
    }

    foreach ($oids as $index => $data) {
        $type = 'apcUPS';
        $descr = 'Current Drawn';

        if (isset($data['upsHighPrecOutputCurrent'])) {
            $current_oid = '.1.3.6.1.4.1.318.1.1.1.4.3.4.' . $index;
            $divisor = 10;
            $current = $data['upsHighPrecOutputCurrent'] / $divisor;
        } else {
            $current_oid = '.1.3.6.1.4.1.318.1.1.1.4.2.4.' . $index;
            $divisor = 1;
            $current = $data['upsAdvOutputCurrent'];
        }

        discover_sensor($valid['sensor'], 'current', $device, $current_oid, $index, $type, $descr, $divisor, 1, null, null, null, null, $current);
    }
```

File: includes/discovery/sensors/load/apc.inc.php [hunk 1]
```diff
@@ -4,39 +4,51 @@
...
-            $item['divisor'] = 1;
-        } else {
-            $current_oid = '.1.3.6.1.4.1.318.1.1.1.4.3.3';
-            $value = explode(" ", $oids);
-            $current = $value[1]/$item['divisor'];
+    $three_phase = snmp_get($device, 'upsBasicOutputPhase.0', '-OsqvU', 'PowerNet-MIB');
+    echo $three_phase;
+    if ($three_phase == 3) {
+        for ($ph_ld = 1; $ph_ld <= 3;) { // Input Current 3Phase
+            $ph_load = snmp_get($device, 'upsPhaseOutputPercentLoad.1.1.'.$ph_ld, '-OsqvU', 'PowerNet-MIB');
```
Comment (laf) [inline includes/discovery/sensors/load/apc.inc.php]: Same again with multi get

File: includes/discovery/sensors/load/apc.inc.php [hunk 2]
```diff
@@ -4,39 +4,51 @@
...
+                'index'       => 0,
+                'descr'       => 'Load',
+                'divisor'     => 10,
+                'mib'         => '+PowerNet-MIB',
+            ),
+        );
+        foreach ($oid_array as $item) {
+            $oids           = snmp_get($device, $item['HighPrecOid'].'.'.$item['index'], '-OsqnU', $item['mib']);
+            if (empty($oids)) {
+                $oids        = snmp_get($device, $item['AdvOid'].'.'.$item['index'], '-OsqnU', $item['mib']);
```
Comment (laf) [inline includes/discovery/sensors/load/apc.inc.php]: Same again with multi get

File: includes/discovery/sensors/voltages/apc.inc.php [hunk 3]
```diff
@@ -2,89 +2,121 @@
 
 // APC Voltages
 if ($device['os'] == 'apc') {
-    $oids = snmp_walk($device, '.1.3.6.1.4.1.318.1.1.8.5.3.3.1.3', '-OsqnU');
-    d_echo($oids."\n");
+    $three_phase = snmp_get($device, 'upsBasicOutputPhase.0', '-OsqvU', 'PowerNet-MIB');
+    if ($three_phase == 3) {
+        $phase_divisor = 1;
+        for ($ph_vlin = 1; $ph_vlin <= 3;) {
+            $phinvoltage = snmp_get($device, 'upsPhaseOutputVoltage.1.1.'.$ph_vlin, '-OsqvU', 'PowerNet-MIB');
```
Comment (laf) [inline includes/discovery/sensors/voltages/apc.inc.php]: Same here with multi get

File: includes/discovery/sensors/voltages/apc.inc.php [hunk 4]
```diff
@@ -2,89 +2,121 @@
...
+            $phvlin_oid = '.1.3.6.1.4.1.318.1.1.1.9.2.3.1.3.1.1';
+            discover_sensor($valid['sensor'], 'voltage', $device, $phvlin_oid.'.'.$ph_vlin, $phvlin_oid.'.'.$ph_vlin, apc, 'Phase '. $ph_vlin .' Input Voltage', $phase_divisor, 1, null, null, null, null, $phinvoltage);
+            $ph_vlin++;
+        }
+        for ($ph_vlout = 1; $ph_vlout <= 3;) {
+            $phoutvoltage = snmp_get($device, 'upsPhaseInputVoltage.1.1.'.$ph_vlout, '-OsqvU', 'PowerNet-MIB');
```
Comment (laf) [inline includes/discovery/sensors/voltages/apc.inc.php]: Same here with multi get

**Comment by NerdBlender** [comment]
@murrant - that doesn't work - as upsHighPrecOutputCurrent is always present - so $oids will always be populated - but i think i can use that to make something that works.

**Comment by laf** [comment]
I've asked the people in the issues if they can test this.

**Comment by wriedel** [comment]
Hi Neil,
many thanks for picking this up. Currently it's not really working for me. I have two apc-surt-20000 and one is now showing voltage for two phases while the other still for one while in reality I have three phases

here some debugs:
http://pastebin.com/0KDGyugw
http://pastebin.com/eYuwv5Vi
and SNMP walks
http://pastebin.com/PuT83sTP
http://pastebin.com/5tVnjQuR

I am not sure on dependencies on #5626 as this bug broke my discovery process.

Thank you,
Wolfgang

**Comment by NerdBlender** [comment]
@wriedel - Can you get me the output of /usr/bin/snmpbulkwalk -v2c -c 'YOUR COMMUNITY' -OQUs -m IF-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/apc udp:apc-surt-20000-b.toocoolforyou.net:161  1.3.6.1.4.1.318

Your snmpwalk is missing the APC info

**Comment by NerdBlender** [comment]
@wriedel - Try this latest version - i have changed to another OID to use for detection.

**Comment by NerdBlender** [comment]
Sorry, that should have been:

 /usr/bin/snmpbulkwalk -v2c -c 'YOUR COMMUNITY' -OQUs -m PowerNet-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/apc udp:apc-surt-20000-b.toocoolforyou.net:161 1.3.6.1.4.1.318

However the latest update I did this morning make the code read the number of phases correctly.

**Comment by wriedel** [comment]
OK, here we are: http://pastebin.com/KcLPhCSY
To actually test it I need to do this, right?
```
gitpull=5558
cd /opt/librenms/
wget https://patch-diff.githubusercontent.com/raw/librenms/librenms/pull/$gitpull.diff
git apply $gitpull.diff
```

**Comment by NerdBlender** [comment]
I'm not sure about the diff pull - i'm not a git expert @laf might suggest the best method.

However, I think I know why this is wrong. This UPS can be configured with 3phase in and 3phase out, OR with 3phase in and a single phase output which looks to be how yours is configured, so i will need to add some extra logic to check for the number of output phases.

**Comment by wriedel** [comment]
Yes, that is correct. 
In fact you can hard-wire 1 or 3 Phase in and 1 or 3 Phase out.
Currently I have 3 Phase in and 1 Phase out as there is a dependency on 1 Phase per battery pack and we just have one battery pack per UPS.
On Cacti I just pulled all 3 Phases in and out all the time as if unused they're just zero.

**Comment by NerdBlender** [comment]
@wriedel  - this should work for you now, i managed to test with someone with a similar setup, although a different model.

**Comment by wriedel** [comment]
@NerdBlender good Morning,
is this change already in the current diff for 5558?
If so, sorry still no luck but not sure if I need to delete and add the device again?

After applying the patch I see this
 	2017-01-28 11:09:52 	Sensor Deleted: voltage apc 3.2.1.0 Input 	
	2017-01-28 11:09:52 	Sensor Deleted: voltage apc 4.2.1.0 Output 	
	2017-01-28 11:09:51 	Sensor Added: voltage apcIn 0 Input 	
	2017-01-28 11:09:51 	Sensor Added: voltage apcOut 0 Output
but still just one graph.

Attached the current debugs for discovery and poller.
Wolfgang

[apc-surt-20000-a.toocoolforyou.net.log.txt](https://github.com/librenms/librenms/files/736851/apc-surt-20000-a.toocoolforyou.net.log.txt)

**Comment by NerdBlender** [comment]
could you get me the output of /usr/bin/snmpbulkwalk -v2c -c 'YOUR COMMUNITY' -OQUs -m PowerNet-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/apc udp:apc-surt-20000-b.toocoolforyou.net:161 1.3.6.1.4.1.318

Pastebin removed the last one.

**Comment by wriedel** [comment]
@NerdBlender 
here we go: http://pastebin.com/JANzu3af

Many thanks!!!
Wolfgang

**Comment by NerdBlender** [comment]
@wriedel - Just pushed another change, let me know if this solves it for you. If it doesn't - can you upload a screenshot of what you are seeing, might help understand whats missing.

**Comment by wriedel** [comment]
@NerdBlender Looks much better now, seems we're getting there ;-)

I had to remove both UPS from the inventory as I ended up with a mess but now lit looks good for the Voltage just something with the current is still screwed. 

Current
Phase 1 Output 		3.5A
Phase 1 Input 		12A
Phase 2 Input 		-0.1A
Phase 3 Input 		-0.1A
I attach two Cacti as also the LibreNMS screen shots.

What also would be very useful if you could expand the apc.yaml 
over:
    - { graph: device_current, text: Current }
to also include the summary for Voltage as also the run-time.

Thank you so much!
Wolfgang

librenms-3phase-current
apc-3phase-voltage
apc-3phase-current

**Comment by laf** [comment]
Let us know when you feel this is ready.

**Comment by NerdBlender** [comment]
@wriedel - this should all be good now.

**Comment by wriedel** [comment]
Sorry for being a pain but we're really close ;-)
Now Current for "Phase x Output" is swapped with "Phase x Input"
screen shot 2017-02-02 at 14 31 38pm
screen shot 2017-02-02 at 14 34 54pm

and here what APC tells me:
screen shot 2017-02-02 at 14 33 21pm

**Comment by NerdBlender** [comment]
With any luck, this should be good now. I fixed a couple of other things, so it wont show the phase if the value is -1

**Comment by wriedel** [comment]
@NerdBlender now it looks perfect!!!!
Everything is as it should be and the quick view with runtime, voltage and current  is ideal to get a brief overview!

Well done and again, many thanks!
Wolfgang

@laf now I really feel we're ready ;-)

**Comment by laf** [comment]
Has anyone tested this branch against non-3 phrase APC units to confirm it still works.

**Comment by wriedel** [comment]
Hi @laf ,
I can setup something with singe phase but I can't pull the patch anymore?
Not sure what I am doing wrong?

What I do:
gitpull=5558
cd /opt/librenms/
wget https://patch-diff.githubusercontent.com/raw/librenms/librenms/pull/$gitpull.diff
git apply $gitpull.diff

What I get:
error: patch failed: includes/discovery/sensors/load/apc.inc.php:4
error: includes/discovery/sensors/load/apc.inc.php: patch does not apply
for every file...

**Comment by laf** [comment]
The current branch needs rebasing against out master (we tidied up sensor discovery).

@NerdBlender can you do this?

**Comment by wriedel** [comment]
@NerdBlender @laf is there anything I can do on my side to get this finialized?

**Comment by murrant** [comment]
You'll need to rebase your changes so they can be applied against master.

http://docs.librenms.org/Developing/Using-Git/

**Comment by NEO-AMiGA** [comment]
From the looks of it something have changed over time and I've lost many of the sensors in our UPS. See #2733 .


This is what it looks like here now:
smartups03

This is what another monitoring solution gives me, and what I'd expect when working 100%.
smartups_other_app


Cheers and thanks

**Comment by wriedel** [comment]
@NEO-AMiGA  I have the same issue but it's fixed with #5558

**Comment by wriedel** [comment]
@murrant, @laf, @NerdBlender 
I am really not a software developer and at the current point I don't feel comfortable to play around. Not sure what's up with @NerdBlender  but maybe someone else can take over the branch and do the rebase dance?
Wolfgang

**Comment by laf** [comment]
I'm sure @NerdBlender will be back to rebase this soon

**Comment by NerdBlender** [comment]
Hi Folks - I was out last week, it will try and rebase this week if i can.

**Comment by wriedel** [comment]
Hi Folks,
just wonder if there is anything I can help with to get this re-base thing done so we can close this one and get it out?
Thank you,
Wolfgang

**Comment by laf** [comment]
I've pushed out a notification to ask APC users to test this so we can ensure it doesn't break anything: https://community.librenms.org/t/looking-for-apc-users-to-test-an-update/654

**Comment by wriedel** [comment]
Hi @laf,
re-base dance again ;-) 
It's broken since yesterday, after the big schema update but worked like a charm before.

git checkout issue-2733
Switched to branch 'issue-2733'
Your branch is ahead of 'NerdBlender/issue-2733' by 15 commits.

root@thor-odd:/opt/librenms# ./validate.php 
==========================================================
Component | Version
--------- | -------
LibreNMS  | 294c799bc8ade001fe0365373b39b37e4606b42f
DB Schema | 171
PHP       | 7.0.15-1
MySQL     | 10.2.4-MariaDB-10.2.4+maria~sid-log
RRDTool   | 1.6.0
SNMP      | NET-SNMP 5.7.3
==========================================================

**Comment by NerdBlender** [comment]
Rebased.

**Comment by wriedel** [comment]
Hi Folks,
I have verified proper functionality for 3-Phase Smart-UPS RT 20000 RM XL (AOS v6.2.1 / App v6.2.1) and 1-Phase Smart-UPS RT 5000 XL 5.1 (AOS v6.4.0 / App v6.4.0) and everything looks good on my end running this now over one week in production. I think we're good to go!

**Comment by laf** [comment]
@NerdBlender Looks like `.1.3.6.1.4.1.318.1.1.1.4.2.4` is no longer in this new code, how come it was taken out?

We can also shift the phase count into pre-cache, I'll sort that out unless you get to it.

**Comment by NerdBlender** [comment]
Took out .1.3.6.1.4.1.318.1.1.1.4.2.4 (upsAdvOutputCurrent) as its kinda sketchy on three phase - it seems to be hit an miss which models support this - in favour of 1.3.6.1.4.1.318.1.1.1.4.3.4 (upsHighPrecOutputCurrent) which seems to work on all models. If its an issue i can alter it.

**Comment by laf** [comment]
The issue is that data will now disappear for users that have it now. That's not an issue if you think this is genuinely the best way forward, it just means we will have to notify users ahead of this PR.

**Comment by wriedel** [comment]
Hi Folks,
sorry really don't want to open another can of worms but based on a escalation we had today it looks vital to monitor "Actual Battery Bus Voltage". So if you have a chance to add this sensor would really be appreciated from a lot of APC customers I guess ;-)

upsAdvBatteryActualVoltage OBJECT-TYPE
  SYNTAX INTEGER
  ACCESS read-only
  STATUS mandatory
  DESCRIPTION
     "The actual battery bus voltage in Volts."
  ::= { upsAdvBattery 8 }

**Comment by NerdBlender** [comment]
As as far as I can work out - upsAdvOutputCurrent/voltage has a precision of 1, where upsHighPrecisionOutputCurrent/voltage has a precision of 0.1. I think this is probably what we want.

**Comment by NerdBlender** [comment]
Added battery voltage and moved phasecount to pre cache.

**Comment by wriedel** [comment]
I never thought that I would have been exited over UPS monitoring but this really rocks!
@NerdBlender many thanks for getting the "Battery Bus Voltage" in. Works like a charm!
Thank you,
Wolfgang

**Comment by wriedel** [comment]
@laf and @NerdBlender one more thing but I am not sure if this is related to APC UPS or more a general problem. We have a problematic UPS which is stuck at a Battery Charge of 99,9%.
While this is correct:
screen shot 2017-03-10 at 15 00 15pm
I have here 100% and I am not sure where this math problem is coming from?
screen shot 2017-03-10 at 15 00 29pm

My point is that 99,9 is not equal 100 and an UPS running at a level no equal to 100 will never start a self test. So my question is if it makes sense to be a little bit more granular while rounding up numbers?

**Comment by NerdBlender** [comment]
@wriedel Thanks for the feedback. Hopefully this is all good now and can be merged in when ready.

As for the issue, I wonder if its a difference between the precision of the discovery and how the RRD's are created, we probably should create a separate issue for that.

**Comment by murrant** [comment]
The rounding is only for the graph, the data in the database (for alerting) and rrd is accurate.

**Comment by kakopedreros** [comment]
Done!

The UPS is a APC Smart-UPS RT 10000 XL 3 Phase UPS with 3-phase input and 1-phase output.

Doesn't shows the 3-phase input, or the 1-phase output.

I checkout the issue-5558 from the NerdBlender repository.

The SNMP walk is this:  http://pastebin.com/1WJeFF1J

And the this a screenshoot of the sensors:
captura de pantalla 2017-03-11 a las 20 12 12

And this are a screenshoot of the resent events:

captura de pantalla 2017-03-11 a las 20 13 26

**Comment by NerdBlender** [comment]
@kakopedreros - can you get me the output of:

/usr/bin/snmpbulkwalk -v2c -c 'YOUR COMMUNITY' -OQUs -m PowerNet-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/apc udp:apc-surt-20000-b.toocoolforyou.net:161 1.3.6.1.4.1.318 <your ups name>

**Comment by NerdBlender** [comment]
@kakopedreros @wriedel 

Can you test with this latest patch - I have changed the OID we detect the three phase on.

**Comment by wriedel** [comment]
@NerdBlender everything looks good on Smart-UPS RT 20000 RM XL and Smart-UPS RT 5000 XL

**Comment by kakopedreros** [comment]
@NerdBlender here is the snmpbulkwalk requested.

[snmpwalk_dump.txt](https://github.com/librenms/librenms/files/840113/snmpwalk_dump.txt)

**Comment by kakopedreros** [comment]
After applied the lastest patch:

[root@librenms librenms]# git pull NerdBlender issue-2733
From https://github.com/NerdBlender/librenms
 * branch            issue-2733 -> FETCH_HEAD
Updating 27c4d56..ac914af
....

Only shows one output phase, not the 3-phase input.

Model: Smart-UPS RT 10000 RM XL (Now full charged and load).

**Comment by NerdBlender** [comment]
@kakopedreros 

Can you run this

/usr/bin/snmpbulkwalk -v2c -c 'YOUR COMMUNITY' -OQUs -m PowerNet-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/apc udp:<your UPS hostname>:161 1.3.6.1.4.1.318

Make sure you include the 1.3.6.1.4.1.318 at the end - apologies if that is what you already put in the dump, but it seems to be missing some info, or your UPS firmware is weird.

**Comment by kakopedreros** [comment]
@NerdBlender here is the snmpbulkwalk with the executed command... =)

[snmpbulkwalk_2.txt](https://github.com/librenms/librenms/files/842786/snmpbulkwalk_2.txt)

**Comment by NerdBlender** [comment]
@kakopedreros - can you run 

 /opt/librenms/discovery.php -d -v -h hostname -m sensors 

(i presume your librenms install is in /opt)

That SNMP Walk doesn't appear right somehow - there seems to be a lot of SNMP data missing compared to mine or @wriedel's output. @wriedel - what firmware version are you using on your SmartUPS RT 20000 XL?

**Comment by wriedel** [comment]
Hi @NerdBlender ,
we have:

Smart-UPS RT 20000 RM XL
APC Management Module AOS v6.2.1 / App v6.2.1 

Smart-UPS RT 5000 XL 5.1 (ID5)
APC Management Module AOS v6.4.0 / App v6.4.0.

We should have a SURT 8000 somewhere which I may be able to bring up if needed.
If I am not mistaken we used the following OID's on Cacti for SURT 8000 / SURT 10000:
APCInputLineCurL0 -> .1.3.6.1.4.1.318.1.1.1.4.2.4.0
APCInputLineVAL0 -> .1.3.6.1.4.1.318.1.1.1.4.2.3.0
APCInputLineVolt0 -> .1.3.6.1.4.1.318.1.1.1.3.2.1.0
APCOuttLineCurL0 -> .1.3.6.1.4.1.318.1.1.1.4.2.4.0
APCOutputLineVolt0 -> .1.3.6.1.4.1.318.1.1.1.4.2.1.0

**Comment by murrant** [comment]
What is the current status of this?

**Comment by NerdBlender** [comment]
@murrant - i am waiting for @kakopedreros to give me the info requested so that we can ensure something isn't broken - i cannot replicate what he is seeing on any of my UPS's, and @wriedel cannot either, so I don't know if its a bug in the code, or a bug in his UPS firmware or what.

**Comment by kakopedreros** [comment]
Here is the requested command:
[sensors.txt](https://github.com/librenms/librenms/files/869263/sensors.txt)

The firmware version on my UPS:
**Smart-UPS RT 10000 RM XL 476.19.W
APC Management Module AOS v6.2.1 / App v6.2.1** 

captura de pantalla 2017-03-24 a las 18 33 19

captura de pantalla 2017-03-24 a las 18 34 01

**Comment by NerdBlender** [comment]
@kakopedreros -

can you post me a screenshot of your UPS overview page direct from the management card?

**Comment by kakopedreros** [comment]
@NerdBlender this one?

captura de pantalla 2017-03-30 a las 23 06 59

**Comment by kakopedreros** [comment]
The Datasheet:

http://www.apc.com/shop/my/en/products/APC-Smart-UPS-RT-10000VA-230V/P-SURT10000XLI

**Comment by NerdBlender** [comment]
Thanks @kakopedreros 

Are you certain that this is wired as a 3 phase UPS?

image

Mines older firmware, but it should show the phases on that overview page

What does it show on the status page. To me it looks like its either wired up as single phase, or somehow the management card does not know its a three phase UPS somehow.

@wriedel - could you post the same page from one of yours, as they are similar models.

**Comment by kakopedreros** [comment]
@NerdBlender maybe is what you say. The management card doesn't recognize it is a 3-phase UPS...

A pic of the INPUT connectors of the UPS...

img_0816

**Comment by NerdBlender** [comment]
@laf @murrant any ideas if there is anything else that can be done to work out where the issue is?

The SNMP walk seem to be missing the OIDs that handle 3 phase UPS data output. Ever seen anything like that before?

**Comment by wriedel** [comment]
@NerdBlender if my memory serve me right, it's implemented differently on these older UPS which didn't have the 3 Phase measurement capabilities and therefore the OID seem to be different:

Input Current:  .1.3.6.1.4.1.318.1.1.1.4.2.4.0
Input Power: .1.3.6.1.4.1.318.1.1.1.4.2.3.0
Input Voltage: .1.3.6.1.4.1.318.1.1.1.3.2.1.0
Output Current : .1.3.6.1.4.1.318.1.1.1.4.2.4.0
Output Voltage: .1.3.6.1.4.1.318.1.1.1.4.2.1.0

Everything else should be similar I guess.
Wolfgang

**Comment by laf** [comment]
@wriedel I guess if there are oid's available for the older models then you can use those, if we record models / versions in polling you could also base the detection on that although imho that's messy.

**Comment by wriedel** [comment]
@laf @NerdBlender @kakopedreros @murrant my point is that SURT 8000 / SURT 10000 / and so on may have 3Phase connectivity but even inside APC didn't really count as 3 Phase UPS. 
Different product line and support team!
So it would make sense to put them into the non 3Phase bucket and use the other OID's.

**Comment by laf** [comment]
I'll trust what you say :)

Our goal is not to lose data so if what you propose doesn't do that and is the correct choice then go ahead.

**Comment by f0o** [comment]
@laf is your changes request still valid? I couldnt find your points in the recent files

**Comment by laf** [comment]
Seems ok, if @NerdBlender can just confirm from my last comments we should be good to go.

**Comment by laf** [comment]
@NerdBlender bump

**Comment by NerdBlender** [comment]
@laf - I am good with your comments.

Let me know if you need anything from me.

**Comment by wriedel** [comment]
Hi Folks, just wonder what's missing so we can finally close this PR and get it out of the door ;-)

**Comment by laf** [comment]
I guess just the conflict to fix + sign the new CLA

**Comment by wriedel** [comment]
@NerdBlender 

yes, too bad as this now breaks the current test setup..

git pull NerdBlender issue-2733
From https://github.com/NerdBlender/librenms
 * branch                issue-2733 -> FETCH_HEAD
Auto-merging includes/discovery/sensors/pre-cache/apc.inc.php
CONFLICT (content): Merge conflict in includes/discovery/sensors/pre-cache/apc.inc.php
Automatic merge failed; fix conflicts and then commit the result.

./scripts/github-apply 5558
also created tons of errors..

**Comment by laf** [comment]
@NerdBlender I've rebased this now, if you can just sign the new CLA: https://cla-assistant.io/librenms/librenms?pullRequest=5558 then we can get it merged.

**Comment by NerdBlender** [comment]
Hi All - Apologies, I have been travelling and haven't had opportunity to catch up with this. Thanks @laf for re-basing, I have signed the agreement

**Comment by laf** [comment]
I've pushed some small code formatting but also the type from apcUPS to apc to keep things consistent.

@NerdBlender can you double check this is ok.

**Comment by NerdBlender** [comment]
@laf all looks good to me.

**Pull request merged**